artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java

```diff
@@ -1249,6 +1256,14 @@ public RoutingStatus send(final ServerMessage message, final boolean direct) thr
 
    @Override
    public RoutingStatus send(final ServerMessage message, final boolean direct, boolean noAutoCreateQueue) throws Exception {
+
+      // If the protocol doesn't support flow control, we have no choice other than fail the communication
+      if (!this.getRemotingConnection().isSupportsFlowControl() && pagingManager.isDiskFull()) {
+         ActiveMQIOErrorException exception = ActiveMQMessageBundle.BUNDLE.diskBeyondLimit();
+         this.getRemotingConnection().fail(exception);
+         throw exception;
```

Comment:
When I say QoS I mean the QoS sent by the client.  In the MQTT case, if you
send a message with QoS1 to the broker, the broker must ack it, otherwise
the client assumes that the message was never received then retries at some
later date.  So it's not a silent fail.  The client never thinks the broker
got the message.

With AMQP, STOMP and OpenWire.  It's a little harder than just blocking
using flow control, there's no way (at least that I know of) to revoke
credits once they've been allocated in AMQP and CORE.  This means clients
are able to continue sending potentially causing an OOM.

On Mon, Sep 5, 2016 at 3:00 PM, clebertsuconic <notifications@github.com>
wrote:

> In artemis-server/src/main/java/org/apache/activemq/artemis/
> core/server/impl/ServerSessionImpl.java
> <https://github.com/apache/activemq-artemis/pull/749#discussion_r77526135>
> :
>
> > @@ -1249,6 +1256,14 @@ public RoutingStatus send(final ServerMessage message, final boolean direct) thr
> >
> >     @Override
> >     public RoutingStatus send(final ServerMessage message, final boolean direct, boolean noAutoCreateQueue) throws Exception {
> > +
> > +      // If the protocol doesn't support flow control, we have no choice other than fail the communication
> > +      if (!this.getRemotingConnection().isSupportsFlowControl() && pagingManager.isDiskFull()) {
> > +         ActiveMQIOErrorException exception = ActiveMQMessageBundle.BUNDLE.diskBeyondLimit();
> > +         this.getRemotingConnection().fail(exception);
> > +         throw exception;
>
> Failng the send here would endup the STOMP or MQTT producer to believe the
> message was sent.
>
> if you want that level of QOS set global-max-size to something meaninful
> and configure it to DROP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/activemq-artemis/pull/749/files/f84d9d105901253d079ec8dd42b888f15f356354#r77526135>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOCy9VkkrKDq5uWaBYyXKvTfgU4K6qTks5qnCBsgaJpZM4J0MjB>
> .
>